### DISCUSSION TITLE
Specify username for Redis authentication

### DISCUSSION THREAD

**Discussion by game5413** [laravel/framework#38737] [Q&A] [answered]
I was wondering, can laravel specify Redis user that want authenticated from `.env` ? i'm using phpredis and lookup into documentation [link](https://github.com/phpredis/phpredis#auth) it can be specified using array or array associative. Maybe something like `REDIS_USERNAME` on `.env` side, i'd tried that already, it seem that `illuminate/redis` only using password for  authentication `$client->auth($config['password']);`.

**Comment by xelab** [comment, marked as answer]
Hello, I'm waiting this feature too but FYI there is a workaround, in your database.php file you can set this : 
`'redis' => [
     'default' => [
            ...,
            'password' => ['user' => 'user', 'pass' => 'pass'],
            ...
        ],
]`

**Reply by barismar** [reply]
Thanks!
This is a workaround but I found a better way to do that, you can add context in your database config **starting from Laravel 6**.
<img width="289" alt="Screen Shot 2021-11-08 at 14 20 33" src="https://user-images.githubusercontent.com/36620482/140699889-92b3cc2d-51c2-4706-ba46-419b915f2238.png">
It only works if the version of your **phpredis extension is >= 5.3.0**, to check the version please go to this [Link](https://github.com/phpredis/phpredis/issues/2#issuecomment-942780).

**Reply by xelab** [reply]
You're right, that's a more elegant way!

**Comment by game5413** [comment]
@xelab I'm already tried that, but still the user that used for authenticated still not changed. My first schema create new user with name `admin` for administration,  and create another user with name `myuser` for general task, and deleting the `default` user. After set config from database.php still not changed (and got an error because user `default` already deleted), i think because Laravel not passing username like the [example ](https://github.com/phpredis/phpredis#example-4) from phpredis. It only using `SETNAME` for changing display name on `CLIENT LIST`, so not changing the user name that we using for authentication. In below i'm attaching the `illuminate/redis` code when creating connection.

![image](https://user-images.githubusercontent.com/37659721/140056737-ba625193-9b79-463b-8dfe-74afff0e1115.png)

Because the first schema was failed, for now i'm just create username `admin` for administration and login by that user and change for permission the `default` user.

**Reply by xelab** [reply]
That's weird, I have no problem with my solution, as I put in "$config['password'] the whole associative array with user and pass, I guess params are passed like PHPRedis wants it. It's important to put an associative array, I tried with a simple array and that didn't work.

**Reply by game5413** [reply]
Ahh.. i see, i'm trying with just an regular array back then, but i don't think that so selective before because the example provided the regular array method and associative array. I not try yet using associative array, but you already did. Thanks for the information @xelab .

**Reply by game5413** [reply]
And yeah, it just like weird if we put an array into password it seem like have two password 😅

**Reply by xelab** [reply]
Yes, that's just a workaround :sweat_smile: I agree with you that associative array or not should not matter (maybe it was another mistake for me, I'll retry some day). Btw I'm sorry, I don't understand why that's doesn't work for you (I use php 8 and Laravel 8 FYI).

**Reply by game5413** [reply]
It's ok, thanks to you there was workaround for these (i can't call these a issue because i don't know it was created like that with a purpose or not), and maybe there was a mistake from me when using array. I'm still using PHP 7.4 and Lumen 8 (I think because my Lumen dependecies still using early old 8 version, but i realy don't know for sure).

**Reply by xelab** [reply]
I'm not sure, but maybe your phpredis version is not the latest and doesn't implement the auth with user/pass, that could explain why my workaround doesn't work for you. Authentication with user/pass (ACL) on Redis is quite new (since version 6).

**Reply by xelab** [reply]
Oh, I didn't understand in your last message that's finally worked, good news!

**Comment by sudoanand** [comment]
You can use `php-redis` to authenticate using ACL (username and password combo)

Following example works for me using AWS MemoryDB for redis.

```
$redis = new \Redis();
$redis->connect('tsl://ENDPOINT', 6379);
$redis->rawCommand("auth", "USERNAME", "PASSWORD");
echo $redis->set("test","2");
echo $redis->get("test");
```

**Comment by jyxjjj** [comment]
IDK what version of laravel this issue has been started but why don't you thought about to pass "name:pass" in password field? AUTH command is supported by both space and ":". So... Or maybe it is a feature published between this issue started and i came here?
